### PR TITLE
add AD engine

### PR DESCRIPTION
**Pull request by Roger-luo**

### CODE REVIEW

**Comment by Roger-luo** [comment]
not sure all the rules we need are forwarded, need to check with the PEPS example and forward whatever is missing

**Comment by GiggleLiu** [comment]
A tried this AD engine a bit. I think the design is good in general. So I will merge it directly. But I do not understand
* Why do we need TrackedArray, TrackedArrayTypes, TrackedRealArray, TrackedComplexArray? Isn't TrackedArray alone enough?
* If we want to support higher order gradients, we need to allow the gradient field being a TrackedArray, now it is forced to have the same type as input.
* We need to support differentiating a Tuple of array, because the AD of OMEinsum is defined on a tuple of array: https://github.com/under-Peter/OMEinsum.jl/blob/d5e22314eec16a6b8f59d9facb497fa9a7f15a0f/src/autodiff.jl#L32

File: lib/Cracker/src/trace.jl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    end
+end
+
+accumulate_grad!(x::Number, grad) = x + grad
+accumulate_grad!(x::AbstractArray, grad) = (x .+= grad)
+
+function backpropagate!(tracked_value, grad)
+    is_tracked(tracked_value) || error("expect tracked value")
+    record = tracked_value.record::Record
+    record.grad = accumulate_grad!(record.grad, grad)
```
Comment (GiggleLiu) [inline lib/Cracker/src/trace.jl:21]: Why is this assignment nessesary?

Comment (Roger-luo) [reply lib/Cracker/src/trace.jl:21]: you need it for scalar case

**Pull request merged**